R/partition.R

```diff
@@ -37,42 +42,87 @@ get_partition_indices <- function(outcomes, training_frac = 0.8, groups = NULL)
-  grps <- unlist(unique(groups))
-  # initialize train groups & set
-  train_grps <- grps
-  train_set <- indices
-  train_set_grp <- groups
-  # initialize fraction of data in train set
-  frac_in_train <- length(train_set) / length(indices)
-  # keep removing data from train set until fraction in train set <= p (e.g. 0.8)
-  while (frac_in_train > p) {
-    # randomly choose a groups
-    grp <- sample(train_grps, size = 1)
-    # remove groups from train groups & set
-    train_grps <- train_grps[train_grps != grp]
-    train_set <- train_set[train_set_grp != grp]
-    train_set_grp <- train_set_grp[train_set_grp != grp]
-    # calcuate fraction of data in train set
+  unique_groups <- unlist(unique(groups))
+  if (is.null(group_partitions)) {
+    # initialize train groups & set
+    train_grps <- unique_groups
+    train_set <- indices
+    train_set_grp <- groups
+    # initialize fraction of data in train set
     frac_in_train <- length(train_set) / length(indices)
+    # keep removing data from train set until fraction in train set <= p (e.g. 0.8)
+    while (frac_in_train > training_frac) {
+      # randomly choose a groups
+      grp <- sample(train_grps, size = 1)
+      # remove groups from train groups & set
+      train_grps <- train_grps[train_grps != grp]
+      train_set <- train_set[train_set_grp != grp]
+      train_set_grp <- train_set_grp[train_set_grp != grp]
+      # calculate fraction of data in train set
+      frac_in_train <- length(train_set) / length(indices)
+    }
+  } else {
+    names_unrecognized <- names(group_partitions[!names(group_partitions) %in% c("train", "test")])
+    if (length(names_unrecognized) > 0) {
+      stop(paste(
+        "Unrecognized name(s) in `group_partitions`:",
+        names_unrecognized
+      ))
+    }
+    in_train_only <- setdiff(group_partitions$train, group_partitions$test)
+    in_test_only <- setdiff(group_partitions$test, group_partitions$train)
+    in_both <- intersect(group_partitions$test, group_partitions$train)
+    in_neither <- setdiff(unique_groups, union(group_partitions$test, group_partitions$train))
+
+    # initialize train & test sets with samples that must be in one or the other
+    train_set <- indices[groups %in% in_train_only]
+    test_set <- indices[groups %in% in_test_only]
+
+    # sample from remaining samples to reach target training fraction
+    remaining <- indices[-c(train_set, test_set)]
+    if (length(remaining) > 0) {
+      num_needed <- round(training_frac * length(indices) - length(train_set))
+      if (num_needed > length(remaining)) {
+        num_needed <- length(remaining)
+      }
+      if (num_needed > 0) {
+        more_train_samples <- indices[sample(remaining, size = num_needed)]
+        train_set <- c(train_set, more_train_samples)
+      }
+    }
   }
-  message(paste0("Fraction of data in the training set: ", frac_in_train, "."))
-  # get train groups indices
+
+  frac_in_train <- length(train_set) / length(indices)
```

Comment:
should we include a warning for certain scenarios that might be unwanted (i.e. training set smaller than test set)?

Reply:
Good suggestion, added here: 6d13ec0